File: UglyNumber/II/UglyNumberII.py
class Solution(object):
    def nthUglyNumber(self, n):
        """
        :type n: int
        :rtype: int
        """
        UgNum = [1 for _ in range(n)]
        t2 = 0
        t3 = 0
        t5 = 0
        for i in range(1, n):
            UgNum[i] = min(UgNum[t2] * 2, min(UgNum[t3] * 3, UgNum[t5] * 5))
            if UgNum[i] == UgNum[t2] * 2:
                t2 += 1
            if UgNum[i] == UgNum[t3] * 3:
                t3 += 1
            if UgNum[i] == UgNum[t5] * 5:
                t5 += 1
        return UgNum[-1]

File: UglyNumber/II/UglyNumberII.cpp
class Solution {
public:
    int nthUglyNumber(int n) {
        vector<int> dp(n, 1);
        int index_2 = 0;
        int index_3 = 0;
        int index_5 = 0;
        for(int i = 1; i < n; i++) {
        	dp[i] = min(min(dp[index_2] * 2, dp[index_3] * 3), dp[index_5] * 5);
        	if(dp[i] == dp[index_2] * 2)
        		index_2++;
        	if(dp[i] == dp[index_3] * 3)
        		index_3++;
        	if(dp[i] == dp[index_5] * 5)
        		index_5++;
        }
        return dp.back();
    }
};